protozfitsreader/__init__.py

```diff
@@ -1,262 +1,313 @@
-        self.numrows = rawzfitsreader.getNumRows()
+    def __iter__(self):
+        return self
 
-    def _extract_field(self, obj, field):
-        # Read a specific field in object 'obj' given as input 'field'
-        if not obj.HasField(field):
-            raise KeyError(str(field))
-        return getattr(obj, field)
+    def __open_events(self):
+        if not self.is_events_table_open:
+            rawzfitsreader.open(self.fname + ":Events")
+            self.is_events_table_open = True
 
     def list_tables(self):
         return rawzfitsreader.listAllTables(self.fname)
 
-    def read_runheader(self):
-        if not self.ttype == "RunHeader":
-            self.ttype = "RunHeader"
-            self._read_file()
-
-        self._read_message()
-        self.header = L0_pb2.CameraRunHeader()
-        self.header.ParseFromString(self.rawmessage)
-
-    def read_event(self):
-        if not self.ttype == "Events":
-            self.ttype = "Events"
-            self._read_file()
-            self.eventnumber = 1
-        else:
-            self.eventnumber += 1
-
-        self._read_message()
-        self.event = L0_pb2.CameraEvent()
-        self.event.ParseFromString(self.rawmessage)
-
     def rewind_table(self):
         # Rewind the current reader. Go to the beginning of the table.
         rawzfitsreader.rewindTable()
 
-    def move_to_next_event(self):
-        # Iterate over events
-        i = 0
-        numrows = i+2
-        # Hook to deal with file with no header (1)
-        if hasattr(self, 'numrows'):
-            numrows = self.numrows
-        # End - Hook to deal with file with no header (1)
-
-        while i < numrows:
-            self.read_event()
-            # Hook to deal with file with no header (2)
-            if hasattr(self, 'numrows'):
-                numrows = self.numrows
-            # End - Hook to deal with file with no header (2)
 
-            # Hook to deal with file with no header (3)
+class Event:
+    def __init__(self, event, run_id, event_id):
+        self.event_id = event_id
+        self.run_id = run_id
+        self._event = event
+
+        _e = self._event                   # just to make lines shorter
+        _w = self._event.hiGain.waveforms  # just to make lines shorter
+
+        self.pixel_ids = toNumPyArray(_w.pixelsIndices)
+        self._sort_ids = np.argsort(self.pixel_ids)
+        self.n_pixels = len(self.pixel_ids)
+        self._samples = toNumPyArray(_w.samples).reshape(self.n_pixels, -1)
+        self.baseline = self.unsorted_baseline[self._sort_ids]
+        self.telescope_id = _e.telescopeID
+        self.event_number = _e.eventNumber
+        self.central_event_gps_time = self.__calc_central_event_gps_time()
+        self.local_time = self.__calc_local_time()
+        self.event_number_array = _e.arrayEvtNum
+        self.camera_event_type = _e.event_type
+        self.array_event_type = _e.eventType
+        self.num_gains = _e.num_gains
```

Comment:
not sure what is the difference between num_gains and num_channels...

Reply:
me neither. I think I openend an issue for this once. ..

Reply:
So looking at the tests, `num_gains` is always zero. 
https://github.com/cta-sst-1m/digicampipe/blob/f28d88083cbeb7c823905a862ec83197040a6aac/digicampipe/tests/test_read_raw_events.py#L145-L152

Reply:
`num_channels` is always -1. There was no test for this value, so I added one, see commit below.

https://github.com/cta-sst-1m/digicampipe/pull/104/commits/fc49186079367072a47420710aff985a4990ebd2

Reply:
ok, @Etienne12345  is this expected? 

why to fields and why they are different. I would expect on field and for DigiCam num_channels = 1

Reply:
ah maybe one of them is supposed to be which channel is used and they other how many channels are available ?

Reply:
in https://github.com/cta-sst-1m/digicampipe/pull/45 @calispac said:
> num_gains is the number of amplification channels per pixel. Some cameras have low and high gain. we only have 1. So num_gains should be 1

But it is zero. 

It can be quickly cross-checked with the rawreader, which I have not written and never touched, so I cannot have messed the rawreader up:
```python
In [1]: from protozfitsreader import rawzfitsreader

In [2]: rawzfitsreader.open('digicampipe/tests/resources/example_10_evts.000.fits.fz:Events')
Out[2]: 0

In [3]: from protozfitsreader.L0_pb2 import CameraEvent

In [4]: event = CameraEvent()

In [5]: event.ParseFromString(rawzfitsreader.readEvent())
Out[5]: 171684

In [6]: event.num_gains
Out[6]: 0

In [7]: event.head.numGainChannels
Out[7]: -1

In [8]: event.foo
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-11-a23058cc480b> in <module>()
----> 1 event.foo

AttributeError: 'CameraEvent' object has no attribute 'foo'
```

The last was just to prove myself, that I cannot have made a typo in the attribute names, typos do not yield zero or minus one, but an AttributeError, exactly as one would expect. Very nice interface.

Reply:
Obviously no. I did not write the python: only the very low-level wrapper around the C++ code. This wrapper only de-serializes protocol buffer structures, any structure (as long as its data model is known). Someone from Barcelona had started with the higher level code (the version which is in ctapipe), while Victor continued, moved it to digicampipe, and heavily reworked it afaik. Most likely he did not care about the number of channels.  

The zfits files have no idea about the number of channels. All they know is whether this or this field is here or not. So most probably the python code should take care of this, by trying to query specific fields, right ?

Reply:
I have a telconf right now, but would it help if I dig in the raw reader if there is anything specific to the number of gains in the low /high level code ?

Reply:
> only the very low-level wrapper around the C++ code.
This is what is imported under the name `rawzfitsreader` correct?

Reply:
Exactly